hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -88,364 +80,131 @@ public String call() {
-                logger.finest("Ignoring member added event " + membershipEvent + " since the client is disconnected.");
-                return;
-            }
-
-            Member member = membershipEvent.getMember();
-            if (members.contains(member)) {
-                logger.finest("Ignoring member added event " + membershipEvent + " since the member is already in the list.");
-                return;
-            }
-
-            logger.finest("New member added to the cluster. Registering " + registrations.size() + " listeners to member "
-                    + member);
-
-            try {
-                getOrConnect(member, client.getClientClusterService().getOwnerConnectionAddress());
-            } catch (Exception e) {
-                logger.warning("Failed to register listeners to member " + member + " rescheduling the registration in "
-                        + SMART_LISTENER_MEMBER_ADDED_RESCHEDULE_TIME + " msecs", e);
-
-                client.getClientExecutionService().schedule(new Runnable() {
-                    @Override
-                    public void run() {
-                        memberAdded(membershipEvent);
-                    }
-                }, SMART_LISTENER_MEMBER_ADDED_RESCHEDULE_TIME, TimeUnit.MILLISECONDS);
-                return;
-            }
-
-            members.add(member);
-
-            for (Map.Entry<ClientRegistrationKey, Map<Member, ClientEventRegistration>> entry : registrations.entrySet()) {
-                ClientRegistrationKey registrationKey = entry.getKey();
-                Map<Member, ClientEventRegistration> registrationMap = entry.getValue();
-                // Only register if not already registered
-                if (null == registrationMap.get(member)) {
+    public void connectionAdded(final Connection connection) {
+        registrationExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                for (ClientRegistrationKey registrationKey : registrations.keySet()) {
                     try {
-                        invoke(registrationKey, member);
+                        invoke(registrationKey, connection);
                     } catch (Exception e) {
-                        logger.warning("Listener " + registrationKey + " can not be added to new member " + member
-                                        + " rescheduling the registration in " + SMART_LISTENER_MEMBER_ADDED_RESCHEDULE_TIME
-                                + " msecs", e);
-                        client.getClientExecutionService().schedule(new Runnable() {
-                            @Override
-                            public void run() {
-                                memberAdded(membershipEvent);
-                            }
-                        }, SMART_LISTENER_MEMBER_ADDED_RESCHEDULE_TIME, TimeUnit.MILLISECONDS);
+                        logger.warning("Listener " + registrationKey + " can not be added to new connection: "
```

Comment:
When connection is disconnected and reconnected, it will be handled in connectionAdded. 
But invocation cancelled because heartbeat timed out is a problem here.  I left that out, because I was thinking it was not handled already. But now looking at the code again, there was a retry. I need to work on this further. 

In your implementation, ClientNonSmartService does not handle this case but ClientSmartService handles it. This is why I got confused. I will try to find a solution for both. And note that all tests are passing. which means there is no test that reproduces that scenario. Given that, it might be better to handle this in a follow-up pr, instead of this one.